Control/Monad/Permutations.hs

```diff
@@ -0,0 +1,220 @@
+-- in: /Parsing Permutation Phrases/, by Arthur Baars, Andres Loh and
+-- Doaitse Swierstra. Published as a functional pearl at the Haskell
+-- Workshop 2001.
+--
+-- From these two works we derive a flexible and general method for
+-- parsing permutations over an Applicative structure. Quite useful
+-- in conjunction with \"Free\" constructions of Applicatives, Monads,
+-- etc.
+--
+-- Other permutation parsing libraries tend towards using special \"almost
+-- applicative\" combinators for construction which denies the library user
+-- the ability to lift and unlift permutation parsing into any Applicative
+-- computational context. We redefine these combinators as convenience
+-- operators here alongside the equivalent Applicative instance.
+--
+-- For example, suppose we want to parse a permutation of:
+-- an optional string of @a@'s, the character @b@ and an optional @c@.
+-- Using a standard parsing library combinator @char@, this can be described
+-- using the Applicative instance by:
+--
+-- > test = runPermutation $
+-- >          (,,) <$> toPermutationWithDefault ""  (some (char 'a'))
+-- >               <*> toPermutation (char 'b')
+-- >               <*> toPermutationWithDefault '_' (char 'c')
+--
+-- Equivalently, this can also be describe using the convenience operators
+-- reminiscent of other parsing libraries:
+--
+-- > test = runPermutation $
+-- >          (,,) <$?> ("", some (char 'a'))
+-- >               <||> char 'b'
+-- >               <|?> ('_', char 'c')
+
+module Control.Monad.Permutations
+  (
+  -- ** Permutation type
+    Permutation()
+  -- ** Permutation evaluators
+  , intercalateEffect
+  , runPermutation
+  -- ** Permutation constructors
+  , toPermutation
+  , toPermutationWithDefault
+  -- ** Convenience operators
+  , (<||>)
+  , (<$$>)
+  , (<|?>)
+  , (<$?>)
+  ) where
+
+import Control.Applicative
+
+-- | An Applicative wrapper-type for constructing permutation parsers.
+
+data Permutation m a = P (Maybe a) (m (Permutation m a))
+
+instance Functor m => Functor (Permutation m) where
+
+    fmap f (P v p) = P (f <$> v) (fmap f <$> p)
+
+instance Alternative m => Applicative (Permutation m) where
+
+    pure value = P (Just value) empty
+
+    lhs@(P f v) <*> rhs@(P g w) = P (f <*> g) (lhsAlt <|> rhsAlt)
+      where
+        lhsAlt = (<*> rhs) <$> v
+        rhsAlt = (lhs <*>) <$> w
+
+-- | \"Unlifts\" a permutation parser into a parser to be evaluated.
+
+runPermutation
+  :: ( Alternative m
+     , Monad m)
+  => Permutation m a -- ^ Permutation specification
+  -> m a             -- ^ Resulting base monad capable of handling the permutation
+runPermutation (P value parser) = optional parser >>= f
+   where
+      f  Nothing = maybe empty pure value
+      f (Just p) = runPermutation p
+
+
+-- | \"Unlifts\" a permutation parser into a parser to be evaluated with an
+-- intercalated effect. Useful for separators between permutation elements.
+---
```

Comment:
This `---` thing breaks the docs. If you actually render the Haddocks, you'll see that only the first paragraph is included. The rest is lost.